moveit_core/planning_scene/src/planning_scene.cpp

```diff
@@ -1807,8 +1807,12 @@ bool PlanningScene::processCollisionObjectRemove(const moveit_msgs::CollisionObj
   else
   {
     world_->removeObject(object.id);
-    removeObjectColor(object.id);
-    removeObjectType(object.id);
+    if (!robot_state_->hasAttachedBody(object.id))
+    {
+      // Don't discard object color and type when an attached object is just being detached
+      removeObjectColor(object.id);
+      removeObjectType(object.id);
+    }
```

Comment:
I don't understand how the patch solves any problem.
`processCollisionObjectRemove` is only triggered for incoming `CollisionObject` messages, but from my perspective the correct way to attach a `CollisionObject` is by sending a single `AttachedCollisionObject` message. [This code](https://github.com/ros-planning/moveit/blob/master/moveit_core/planning_scene/src/planning_scene.cpp#L1559-L1570) removes the object from the world then but does not touch the color/type data - and that's what you want.

So I would argue the actual bug in your use-case is that you send a `REMOVE` message at all.
Still in these lines there is an associated bug, because color and type are removed even when `world_->removeObject(object.id) == false` (which is the case if the object already got attached).

Reply:
yes, right, u attach an obj by sending a `AttachedCollisionObject` with ADD or APPEND action*, but when that happens, we remove the attached `CollisionObject` from the planning scene [here](https://github.com/ros-planning/moveit/blob/f1f23180c41f3b4f9f2194b8f8d6bb407b668b82/moveit_core/planning_scene/src/planning_scene.cpp#L1559). This effectively removes the color (and type) information about the object. Once detached, we add back to the scene [here](https://github.com/ros-planning/moveit/blob/master/moveit_core/planning_scene/src/planning_scene.cpp#L1652), but by then we don't have color or type anymore.
So to my understanding the patch is more or less correct, if not particularly tidy. I can think on better solutions, but all involving much bigger changes. e.g. add the color info to the CollisionObject msg, or handle attach/detach as a different operation than add/remove.

_* offtopic:  what's the difference between APPEND and ADD?_

Reply:
> we remove the attached CollisionObject from the planning scene here

the line you refer to removes the object from the *world inside the planning scene*
afaics the lines in your patch are not called from there.

Reply:
mmm...  :thinking:    I think they are... indirectly;  w/o going deep into the code, `world_->removeObject` notifies observers that we want to `DESTROY` the obj, and the observers will pub a `CollisionObject` with `REMOVE` action.
Otherwise I misinterpreted something

Reply:
> APPEND can add shapes to a collision object. ADD ads new objects

Thanks for the clarification!